engine/src/main/java/net/jqwik/engine/properties/state/ShrinkableChainIteration.java

```diff
@@ -2,11 +2,61 @@
 
 import java.util.*;
 import java.util.function.*;
+import java.util.stream.*;
 
 import net.jqwik.api.*;
 import net.jqwik.api.state.*;
 
 class ShrinkableChainIteration<T> {
+	// Larger values might improve shrink quality, however, they increase the shrink space, so it might increase shrink duration
+	private final static int NUM_SAMPLES_IN_EAGER_CHAIN_SHRINK = Integer.getInteger("jqwik.eagerChainShrinkSamples", 2);
+
+	static class ShrinkableWithEagerValue<T> implements Shrinkable<T> {
+		protected final Shrinkable<T> base;
+		private final ShrinkingDistance distance;
+		final T value;
+
+		ShrinkableWithEagerValue(Shrinkable<T> base) {
+			this.base = base;
+			this.distance = base.distance();
+			this.value = base.value();
+		}
+
+		@Override
+		public T value() {
+			return value;
+		}
+
+		@Override
+		public Stream<Shrinkable<T>> shrink() {
+			return base.shrink();
+		}
+
+		@Override
+		public ShrinkingDistance distance() {
+			return distance;
+		}
+	}
+
+	static class EagerShrinkable<T> extends ShrinkableWithEagerValue<T> {
+		private final List<Shrinkable<T>> shrinkResults;
+
+		EagerShrinkable(Shrinkable<T> base, int numSamples) {
+			super(base);
+			this.shrinkResults =
+				base.shrink()
+					.sorted(Comparator.comparing(Shrinkable::distance))
+					.limit(numSamples)
```

Comment:
This triggers OOM in case `.shrink()` stream is big. See https://github.com/jlink/jqwik/issues/338#issuecomment-1351637147